src/Scrutor/Decoration/DecorationFactory.cs

```diff
@@ -0,0 +1,26 @@
+﻿using System;
+
+namespace Scrutor.Decoration
+{
+    internal static class DecorationFactory
+    {
+        public static Decoration Create(Type serviceType, Type? decoratorType, Func<object, IServiceProvider, object>? decoratorFactory)
```

Comment:
These methods are never/should never be called with both `decoratorType` and `decoratorFactory`, right? I'd like to see these methods split in two - one for each parameter - to avoid passing `null` and if/else checking. Maybe something like `Decorator.WithType` and `Decorator.WithFactory`?